### PR TITLE
[BEAM-7305] Fix issues in and extend documentation for Hazelcast Jet Runner 

### PR DESCRIPTION
**Pull request by jbartok**
Some new features:
* flushing of metrics periodically instead of at job end
* option to make ParDo processors cooperative

Plus some bug fixes and documentation improvements.
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by jbartok** [comment]
R: @mxm

**Comment by jbartok** [comment]
Run Website PreCommit

**Comment by jbartok** [comment]
Ahh, I finally understand why the above check is failing, will fix it a bit later.

File: runners/jet-experimental/build.gradle [hunk 1]
```diff
@@ -42,38 +44,64 @@ dependencies {
...
+    systemProperty "beamTestPipelineOptions", JsonOutput.toJson(["--runner=TestJetRunner"])
 
     classpath = configurations.validatesRunner
     testClassesDirs = files(project(":sdks:java:core").sourceSets.test.output.classesDirs)
     useJUnit {
         includeCategories 'org.apache.beam.sdk.testing.ValidatesRunner'
+        excludeCategories "org.apache.beam.sdk.testing.LargeKeys\$Above100MB"
         excludeCategories 'org.apache.beam.sdk.testing.UsesImpulse' //impulse doesn't cooperate properly with Jet when multiple cluster members are used
         exclude '**/SplittableDoFnTest.class' //Splittable DoFn functionality not yet in the runner
+        exclude '**/ParDoLifecycleTest.class' //some of these tests aren't playing well with our test runner's parallelism...
```
Comment (mxm) [inline runners/jet-experimental/build.gradle]: This looks like an important test to fix.

File: runners/jet-experimental/build.gradle [hunk 2]
```diff
@@ -42,38 +44,64 @@ dependencies {
...
+    description = "Runs tests that require a runner to validate that piplines/transforms work correctly"
+    systemProperty "beamTestPipelineOptions", JsonOutput.toJson(["--runner=TestJetRunner"])
+
+    classpath = configurations.needsRunner
+    testClassesDirs += files(project(":runners:core-construction-java").sourceSets.test.output.classesDirs)
+    testClassesDirs += files(project(":runners:core-java").sourceSets.test.output.classesDirs)
+    testClassesDirs += files(project(":sdks:java:core").sourceSets.test.output.classesDirs)
+    useJUnit {
+        includeCategories "org.apache.beam.sdk.testing.NeedsRunner"
+        excludeCategories "org.apache.beam.sdk.testing.LargeKeys\$Above100MB"
```
Comment (mxm) [inline runners/jet-experimental/build.gradle]: Are all of these passing?

File: runners/jet-experimental/build.gradle (hunk 1 above)
Comment (jbartok) [reply runners/jet-experimental/build.gradle]: I honestly think that in this particular case the tests are the incorrect thing. I think the problem is that the tests make an assumption which doesn't hold in the case of our Runner. They assume that by the time the FIRST `@Teardown` method of a DoFn instance get called, all other instances of the same DoFn have also finished their processing, hence an Exception should already have been thrown. I think this assumption is unnecessarily restrictive, it should be enough to check that by the time the LAST `@Teardown` is being called the Exception has been thrown. I've made changes towards this idea, now the test pass, but pls. let me know if this is not ok.

File: runners/jet-experimental/build.gradle (hunk 2 above)
Comment (jbartok) [reply runners/jet-experimental/build.gradle]: Well, so-so: `697 tests completed, 69 failed, 6 skipped`

**Comment by mxm** [comment]
Could you squash at least the spotless fixup commits? I'll merge then.

**Comment by jbartok** [comment]
Sure, I just reasoned that multiple smaller changes would be simpler to review. Coming up!

File: runners/jet-experimental/src/main/java/org/apache/beam/runners/jet/processors/AbstractParDoP.java [hunk 3]
```diff
@@ -247,7 +255,12 @@ private void processBufferedRegularItems(Inbox inbox) {
 
   @Override
   public boolean tryProcess() {
-    return outputManager.tryFlush();
+    boolean successful = outputManager.tryFlush();
+    if (successful && System.currentTimeMillis() > lastMetricsFlushTime + metricsFlushPeriod) {
```
Comment (viliam-durina) [inline runners/jet-experimental/src/main/java/org/apache/beam/runners/jet/processors/AbstractParDoP.java]: Should use `System.nanoTime`, real-time clock can move back and forth. Also you can use `nextMetricsFlushTime` and save one add operation for most calls.

File: runners/jet-experimental/src/main/java/org/apache/beam/runners/jet/processors/ImpulseP.java [hunk 4]
```diff
@@ -19,28 +19,89 @@
...
+      this.context = context;
+    }
+
+    @Nonnull
+    @Override
+    public Collection<? extends Processor> get(int count) {
+      int indexBase = context.memberIndex() * context.localParallelism();
+      List<Processor> res = new ArrayList<>(count);
+      for (int i = 0; i < count; i++, indexBase++) {
+        res.add(new ImpulseP(indexBase == 0, outputCoder, ownerId));
```
Comment (viliam-durina) [inline runners/jet-experimental/src/main/java/org/apache/beam/runners/jet/processors/ImpulseP.java]: You don't need PMS/PS for this, `Processor.Context` has `globalProcessorIndex()`.

**Pull request merged**